### PR TITLE
WIP VPC SDK

### PR DESCRIPTION
**Pull request by jaredm-ibm**

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,214 @@
+inherit_from: .rubocop_todo.yml
```
Comment (agrare) [inline .rubocop.yml:1]: Deciding on rubocop style should be in a separate PR from adding a new API class, particularly because a lot of unrelated code to adding VPC was changed because of these rubocop rules

File: bin/console [hunk 2]
```diff
@@ -1,14 +1,147 @@
...
+require 'bundler/setup'
+require 'ibm'
+require 'pry'
+require 'thor'
+require 'rest-client'
 
-# You can add fixtures and/or initialization code here to make experimenting
-# with your gem easier. You can also use a different console, if you like.
+# Encapsulate the API entryponts.
+class Sdk
```
Comment (agrare) [inline bin/console:11]: This is definitely interesting, can we investigate this in its own PR?

File: ibm-cloud-sdk.gemspec [hunk 3]
```diff
@@ -1,27 +1,31 @@
...
     `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
   end
-  spec.bindir        = "exe"
+  spec.bindir        = 'exe'
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
-  spec.require_paths = ["lib"]
+  spec.require_paths = ['lib']
 
-  spec.add_dependency "rest-client"
+  spec.add_dependency 'rest-client'
```
Comment (agrare) [inline ibm-cloud-sdk.gemspec:29]: Again all of this formatting change isn't really related

File: lib/ibm/cloud/sdk.rb [hunk 4]
```diff
@@ -1,13 +1,43 @@
...
-require "ibm/cloud/sdk/version"
+require_relative 'sdk/base_service'
+require_relative 'sdk/iam'
+require_relative 'sdk/power_iaas'
+require_relative 'sdk/resource_controller'
+require_relative 'sdk/vpc'
 
 module IBM
+  # Holds the SDK pieces.
+  class CloudSDK
```
Comment (agrare) [inline lib/ibm/cloud/sdk.rb:12]: This should match the file name so `IBM::Cloud::SDK`

File: lib/ibm/cloud/sdk/iam.rb [hunk 5]
```diff
@@ -1,35 +1,36 @@
...
     module SDK
+      # Get a new Bearer using the API Key.
       class IAM < BaseService
         def endpoint
-          "https://iam.cloud.ibm.com".freeze
+          'https://iam.cloud.ibm.com'
         end
 
-        def initialize(api_key)
+        def initialize(api_key) # rubocop:disable Lint/MissingSuper
```
Comment (agrare) [inline lib/ibm/cloud/sdk/iam.rb:12]: Really don't like "linter" directives cluttering code, we should be able to just ignore rubocop when we know it is wrong.  The rules are more like guidelines anyway :laughing:

File: lib/ibm/helpers/class.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# typed: false
+# frozen_string_literal: true
+
+# The base object.
+class Object
+  # Filter out the standard methods and only show the interesting methods.
+  # @return [Array<String>]
+  def im
```
Comment (agrare) [inline lib/ibm/helpers/class.rb:8]: This looks handy for dev but I don't think we want to monkeypatch the base `Object` class

File: lib/ibm/helpers/connection.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require 'time'
+require 'http'
+
+module IBM
+  module Cloud
+    module SDK
+      # Contols tokens.
+      class Connection
```
Comment (agrare) [inline lib/ibm/helpers/connection.rb:10]: Should this be in `lib/ibm/cloud/sdk/connection.rb` ?

**Review by agrare — COMMENTED** [review]
Didn't get too deep into the new VPC classes but overall they look great, a lot of unrelated changes that we should split out into their own PRs

File: .gitignore [hunk 8]
```diff
@@ -10,3 +10,4 @@
 
 # rspec failure tracking
 .rspec_status
+.DS_Store
```
[the remark this answers is not in the document]
Comment (jaredm-ibm) [reply .gitignore:13]: I disagree here. there is no need for storing .DS_Store files ever.

[the remark this answers is not in the document]
Comment (agrare) [reply .gitignore:13]: I agree, but there's also no need for a lot of different files (*.swp, *~, etc...)  If we want to add every file type that we don't committed for every OS/Editor/IDE this list will get enormous.  It is on us as reviewers to make sure that people aren't adding files that they shouldn't as part of the PR review process.

[the remark this answers is not in the document]
Comment (agrare) [reply .gitignore:13]: Actually this just came up in the ManageIQ community recently https://github.com/ManageIQ/manageiq-ui-classic/pull/7305#issuecomment-691489962

[the remark this answers is not in the document]
Comment (jaredm-ibm) [reply .gitignore:13]: This is under the IBM namespace not ManageIQ, which means we can decide our policy for ourselves. I have no problem with a large gitignore file if it keeps the repo clean of cruft and saves from ever having those files checked-in in the first place. Besides I think the handwaving of it'll be large is a bit overblown, we can start with the files we know about now and if someone else comes along with additional requirements we can add those at that time. This really just fixes the problem once and for all.

File: .rubocop.yml (hunk 1 above)
Comment (jaredm-ibm) [reply .rubocop.yml:1]: The rubocop_todo.yml is output from the --auto-gen-config. This file whitelists all of the exceptions found by rubocop so that you can start clean and slowly work through the offences when you have time. If you have your IDE set to lint having all of these previous errors can be overwhelming. This clears your technical debt and ensures that you are not adding new violations making IDE listing more useful.

File: lib/ibm/helpers/connection.rb (hunk 7 above)
Comment (jaredm-ibm) [reply lib/ibm/helpers/connection.rb:10]: Yes, I am still playing with this portion trying to get it right. I'm doing two things in here.
1) Changing from rest-client to http. http looks to be faster than rest-client. https://github.com/httprb/http and in my benchmarks this bears out sort of.
2) Have the token refresh automatically.

File: lib/ibm/helpers/class.rb (hunk 6 above)
Comment (jaredm-ibm) [reply lib/ibm/helpers/class.rb:8]: Yes, it is very handy for dev. It's something I created for a different project that I brought over.  I want this as part of the console experience. It's also very handy for demos and learning. By removing all of the standard methods it easier to discern what the class actually does. We can move it directly into the console file, but then people developing in the cloud plugin won't have it.

File: lib/ibm/cloud/sdk/iam.rb (hunk 5 above)
Comment (jaredm-ibm) [reply lib/ibm/cloud/sdk/iam.rb:12]: This comment disables the linter for this line only. It represents a knowing violation of the linter rule, and is easier to work with when listing in the IDE.

File: lib/ibm/cloud/sdk.rb (hunk 4 above)
Comment (jaredm-ibm) [reply lib/ibm/cloud/sdk.rb:12]: This is a work in progress. I just got it working properly yesterday, and am still evaluating how it functions in practice. I'll clean it up and do it properly when I am satisfied this is the correct solution.

File: .editorconfig [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# EditorConfig is awesome: http://EditorConfig.org
```
[the remark this answers is not in the document]
Comment (jaredm-ibm) [reply .editorconfig:1]: The file could use a once over but the items in there should match the expected spacing for each language. It enforces code consistency and limits people from checking in tab spaced files when the rest of the project/file is space separated. Requiring the plugin is easier and less troublesome than checking each file for consistent spacing in each PR.

File: .gitignore (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaywcarman) [reply .gitignore:13]: I'll weight in here. It may seem scrupulous to reject a small addition to the gitignore, but in general we're going to be better able to maintain this repo if we keep the content strictly relevant to the IBM Cloud SDK. In this case, adding .DS_Store to your global gitignore accomplishes everything you need _and_ gives you autonomy to make changes without community review.

File: bin/console (hunk 2 above)
Comment (jaredm-ibm) [reply bin/console:11]: Sure, how do I split these into their own PRs? Although I created this a couple of weeks ago and with CloudSDK I think it may become redundant.

Comment (agrare) [reply bin/console:11]: I like to split up a branch a couple of different ways depending on how involved the number of commits are.

If you have one or two commits that you want to pull out into a new branch you can do e.g.
```
git checkout -b new_pr_branch master
git cherry-pick sha-from-other-branch
```

If there are a lot of commits you can make a copy of your current branch and `git rebase` to drop unrelated commits.

Comment (agrare) [reply bin/console:11]: I just noticed this pr is off of your fork's master so `git checkout -b new_pr_branch master` won't work as expected.  You can `git checkout -b new_pr_branch upstream/master` where `upstream` is whatever you named your remote for this repo

File: .gitignore (hunk 8 above)
[the remark this answers is not in the document]
Comment (jaredm-ibm) [reply .gitignore:13]: I don't understand this discussion. DS_Store files are never relevant to a repo. It seems backwards to allow someone to check them in and then have a discussion about not checking them in when you can prevent it at the source.

[the remark this answers is not in the document]
Comment (agrare) [reply .gitignore:13]: I don't want to hold up this PR while we debate different gitignore philosophies, adding `.DS_Store` to gitignore isn't related to adding support for the VPC SDK and thus should be a separate PR

File: lib/ibm/helpers/class.rb (hunk 6 above)
Comment (Fryguy) [reply lib/ibm/helpers/class.rb:8]: I'm a heavy :-1: in this in a gem.  Gems should not be patching core classes with their own methods, especially when they are development only.

It might make more sense in a gem like `active_support` or `more_core_extensions` since the purpose of those gems are to patch core classes.

Alternatively, if you are using `pry` you don't need this.  `pry` has the `ls` command which gives you exactly what you are looking for.  See https://gist.github.com/banister/c0fc686ef923c8b87715

File: .editorconfig (hunk 9 above)
[the remark this answers is not in the document]
Comment (agrare) [reply .editorconfig:1]: We prefer to run rubocop on PRs with a bot which is even easier than requiring an editor plugin :smile:

[the remark this answers is not in the document]
Comment (jaredm-ibm) [reply .editorconfig:1]: I don't see how rubocop is relevant. But I give up. I'm redoing everything if you want to run your repo like the miq repos where there are files that are partially space indented and partially tab indented sometimes in the same file. I'm done arguing for a better more user friendly way.
